dev/src/bulk-writer.ts

```diff
@@ -211,25 +224,92 @@ class BulkCommitBatch {
     const stack = Error().stack!;
 
-    return this.writeBatch.bulkCommit().catch(err => {
-      throw wrapError(err, stack);
-    });
+    let originalIndexMap: Map<number, number> = new Map(
+      Array.from(new Array(this.opCount), (_, i) => [i, i])
+    );
+
+    let retryAttempts = 0;
+    while (retryAttempts < MAX_RETRY_ATTEMPTS) {
+      let retryIndexes: number[] = [];
+      await this.backoff.backoffAndWait();
+      try {
+        const results = await this.writeBatch.bulkCommit();
+        retryIndexes = this.processResults(originalIndexMap, results);
+      } catch (err) {
+        retryIndexes = this.processResults(
+          originalIndexMap,
+          [],
+          wrapError(err, stack)
+        );
+      }
+
+      // Map the indexes that are going to be retried back to their
+      // corresponding indexes in the original batch.
+      originalIndexMap = new Map(
+        Array.from(new Array(retryIndexes.length), (_, i) => [
+          i,
+          originalIndexMap.get(retryIndexes[i])!,
+        ])
+      );
+      if (retryIndexes.length > 0) {
+        logger(
+          'BulkWriter.bulkCommit',
+          null,
+          `Current batch failed at retry #${retryAttempts}. Num failures: ` +
+            `${retryIndexes.length}.`
+        );
+
+        this.writeBatch = this.writeBatch._sliceIndexes(retryIndexes);
+      } else {
+        break;
+      }
+      retryAttempts++;
+    }
   }
 
   /**
    * Resolves the individual operations in the batch with the results.
+   *
+   * @param originalIndexMap Map of the results' indexes to the indexes
+   * on resultsMap
+   * @return The indexes of writes that failed with a retryable error.
    */
-  processResults(results: BatchWriteResult[], error?: Error): void {
-    if (error === undefined) {
-      for (let i = 0; i < this.opCount; i++) {
-        this.resultsMap.get(i)!.resolve(results[i]);
-      }
-    } else {
-      for (let i = 0; i < this.opCount; i++) {
-        this.resultsMap.get(i)!.reject(error);
+  processResults(
+    originalIndexMap: Map<number, number>,
+    results: BatchWriteResult[],
+    error?: Error
+  ): number[] {
+    const indexesToRetry: number[] = [];
+    if (error) {
+      results = Array.from({length: originalIndexMap.size}, () => {
+        return {
+          writeTime: null,
+          status: error,
+        };
+      });
+    }
+
+    for (let i = 0; i < results.length; i++) {
+      const writeResult = results[i];
+      const originalIndex = originalIndexMap.get(i)!;
+      if (writeResult.status.code === Status.OK) {
+        this.resultsMap.get(originalIndex)!.resolve(results[i]);
+      } else if (this.shouldRetry(writeResult.status.code)) {
+        indexesToRetry.push(i);
+      } else {
+        this.resultsMap.get(originalIndex)!.reject(writeResult.status);
       }
     }
-    this.completedDeferred.resolve();
+
+    if (indexesToRetry.length === 0) {
+      this.completedDeferred.resolve();
```

Comment:
We also need to resolve this when all retry attempts failed. Please also add a test for this.